tests/websocket.c

```diff
@@ -28,147 +48,412 @@ struct worker_thread {
+		free(wt->cur_hdr_val); /* free the previous header value if there was one */
+		wt->cur_hdr_val_len = length;
+		wt->cur_hdr_val = calloc(length + 1, 1);
+		memcpy(wt->cur_hdr_val, at, length);
+	}
+	wt->debug("%s appended header value data: currently [%.*s]\n", __func__,
+		  (int)wt->cur_hdr_val_len, wt->cur_hdr_val);
+
+	if (wt->cur_hdr_key_len == 20 && strncasecmp(wt->cur_hdr_key, "Sec-WebSocket-Accept", 20) == 0) {
+		free(wt->sec_websocket_accept);
+		wt->sec_websocket_accept = calloc(wt->cur_hdr_val_len + 1, 1);
+		memcpy(wt->sec_websocket_accept, wt->cur_hdr_val, wt->cur_hdr_val_len);
+	}
+
+	wt->hdr_last_cb_was_name = 0;
+	return 0;
+}
+
+
+static int
+ws_on_headers_complete(http_parser *p) {
+	struct worker_thread *wt = p->data;
+	wt->debug("%s (wt=%p)\n", __func__, wt);
+	free(wt->cur_hdr_key);
+	free(wt->cur_hdr_val);
+
+	/* make sure that we received a Sec-WebSocket-Accept header */
+	if (!wt->sec_websocket_accept) {
+		wt->debug("%s: no Sec-WebSocket-Accept header was returned\n", __func__);
+		return 1;
+	}
+
+	/* and that it matches what we expect */
+	int ret = 0;
+	if (strlen(wt->sec_websocket_accept) != wt->ws_response_len
+		|| memcmp(wt->ws_response, wt->sec_websocket_accept, wt->ws_response_len) != 0) {
+		wt->debug("Invalid WS handshake: expected [%.*s], got [%s]\n",
+			(int)wt->ws_response_len, wt->ws_response, wt->sec_websocket_accept);
+		ret = 1;
+	}
+
+	free(wt->sec_websocket_accept);
+	return ret;
+}
+
+static void
+ws_enqueue_frame_for_command(struct worker_thread *wt, char *cmd, size_t sz) {
+	int include_mask = (wt->mask_cfg == MASK_ALWAYS ||
+		(wt->mask_cfg == MASK_ALTERNATE && wt->msg_sent % 2 == 0)) ? 1 : 0;
+
+	unsigned char mask[4];
+	for (int i = 0; include_mask && i < 4; i++) { /* only if mask is needed */
+		mask[i] = rand() & 0xff;
 	}
+	uint8_t len = (uint8_t)(sz); /* (1 << 7) | length. */
+	if (include_mask) {
+		len |= (1 << 7); /* set masking bit ON */
+	} 
+
+	/* apply the mask to the payload */
+	for (size_t i = 0; include_mask && i < sz; i++) {
+		cmd[i] = (cmd[i] ^ mask[i % 4]) & 0xff;
+	}
+	/* 0x81 = 10000001b:
+		1: FIN bit (meaning there's only one message in the frame),
+		0: RSV1 bit (reserved),
+		0: RSV2 bit (reserved),
+		0: RSV3 bit (reserved),
+		0001: text frame */
+	evbuffer_add(wt->wbuffer, "\x81", 1);
+	evbuffer_add(wt->wbuffer, &len, 1);
+	if (include_mask) { /* only include mask in the frame if needed */
+		evbuffer_add(wt->wbuffer, mask, 4);
+	}
+	evbuffer_add(wt->wbuffer, cmd, sz);
+	wt->mask_applied += include_mask;
+}
+
+static void
+ws_enqueue_frame(struct worker_thread *wt) {
+	char ping_command[] = "[\"PING\"]";
+	ws_enqueue_frame_for_command(wt, ping_command, sizeof(ping_command) - 1);
+
+	wait_for_possible_write(wt);
+}
+
+static int
+ws_on_message_complete(http_parser *p) {
+	struct worker_thread *wt = p->data;
+
+	wt->debug("%s (wt=%p), upgrade=%d\n", __func__, wt, p->upgrade);
+	// we've received the full HTTP response now, so we're ready to send frames
```

Comment:
minor, but the style of comments is a bit inconsistent in this file. (`//` vs `/* .. */`)